packages/dataviews/src/dataviews.tsx

```diff
@@ -73,25 +70,23 @@ export default function DataViews< Item >( {
 	paginationInfo,
 	defaultLayouts,
 	selection: selectionProperty,
-	setSelection: setSelectionProperty,
-	onChangeSelection = defaultOnChangeSelection,
+	onChangeSelection,
 }: DataViewsProps< Item > ) {
 	const [ selectionState, setSelectionState ] = useState< string[] >( [] );
 	const isUncontrolled =
-		selectionProperty === undefined || setSelectionProperty === undefined;
+		selectionProperty === undefined || onChangeSelection === undefined;
 	const selection = isUncontrolled ? selectionState : selectionProperty;
-	const setSelection = isUncontrolled
-		? setSelectionState
-		: setSelectionProperty;
 	const [ openedFilter, setOpenedFilter ] = useState< string | null >( null );
 
 	function setSelectionWithChange( value: SelectionOrUpdater ) {
 		const newValue =
 			typeof value === 'function' ? value( selection ) : value;
-		onChangeSelection(
-			data.filter( ( item ) => newValue.includes( getItemId( item ) ) )
-		);
-		return setSelection( value );
+		if ( ! isUncontrolled ) {
+			setSelectionState( newValue );
+		}
+		if ( onChangeSelection ) {
+			onChangeSelection( newValue );
```

Comment:
Can't we just do something like this now:

```js
const setSelectionWithChange = onChangeSelection || setSelectionState
```

Reply:
no we can't, because sometimes you want the selection to be uncontrolled (you don't pass any selection prop) but you want to be notified about selection changes so you provide a `onChangeSelection` prop.

Reply:
When do we want this? Can't we always force it to be controlled if you want to be notified?

Reply:
For me, it's a good idea to support this, just like regular inputs or any raw form element in React. You can pass the value to make it controlled otherwise you can use `onChange` to be notified.

Reply:
Doesn't it log a warning in that case about an uncontrolled value?

Reply:
No, it logs a warning only when you switch from controlled to uncontrolled (or the opposite)